added DatagramSocketEventEmitter abstract class, which UnicastEventEmitt...

kkharma opened the pull request:
...er and MulticastEventEmitter both extend

UnicastEventEmitter and MulticastEventEmitter contained a lot of duplicated code.  also, UnicastEventEmitter did not extend AbstractEventEmitter for some reason, so heartbeats were not possible.  this is an attempt to clean that up.

mohammadkolahdouzan commented on src/main/java/org/lwes/emitter/MulticastEventEmitter.java:
```diff
@@ -44,27 +44,15 @@
-
   /* the multicast interface */
-  private InetAddress iface = null;
+  protected InetAddress iface = null;
```
I think this one can also go to DatagramSocketEventEmitter.

kkharma replied:
actually address in the parent class replaces iface, but i need to make a fix to use it everywhere in this class.

kkharma replied:
looking more closely, this predates my changes.  does anyone know why there there are two InetAddresses (iface and address)?  shouldn't there just be one?

mohammadkolahdouzan replied:
address and iface are actually different, address is the address of the host you will emit the messages to, iface is the network interface (in case you have more than one) on the host that would be used to emit the messages.

kkharma replied:
ah yes, i see that now... given that iface is multicast specific, then i think it should remain this class.

mohammadkolahdouzan commented on src/main/java/org/lwes/emitter/MulticastEventEmitter.java:
```diff
@@ -145,60 +133,10 @@ public int getTimeToLive() {
    */
   @Override
-  public void initialize() throws IOException {
+  protected void createSocket() throws IOException {
```
is the API still backward compatible? If not, should the version change not be a major bump?

kkharma replied:
these changes should preserve backwards compatibility.  initialize is still called as it was previously.

mohammadkolahdouzan commented on src/main/java/org/lwes/emitter/MulticastEventEmitter.java:
```diff
@@ -44,27 +44,15 @@
  * @author Anthony Molinaro
```
now that you are changing this, can you add a method to control loopback too? http://download.java.net/jdk7/archive/b123/docs/api/java/net/MulticastSocket.html#setLoopbackMode(boolean)

kkharma replied:
k will do...

kkharma replied:
mohammad and i discussed and will leave this as is.

kkharma commented:
@mohammadkolahdouzan: updated.

mohammadkolahdouzan commented:
looks glorious to me.

kkharma commented:
not sure what the travis integration stuff is but i'm updating this PR to include changes for more flexible emitter group creation... most of the way there... just need to make a few more changes.

kkharma commented:
this should be good for review again... major changes are creation of an abstract parent class for both MulticastEventEmitter and UnicastEventEmitter... more recent changes include updating the EmitterGroupBuilder parsing logic to handle both unicast and multicast emitters and allowing for specification of interface and ttl for each host.

kkharma commented:
other notes:
1) i removed PreserializedUnicastEventEmitter because i didn't see a purpose for it.
2) also removed a superfluous(?) System.out.println in EmitterGroupBuilder.
3) changed the default TTL to 5 to match the erlang library.

kkharma commented:
latest changes include the addition of a shutdown method for emitter groups

nimaco commented on src/main/java/org/lwes/emitter/DatagramSocketEventEmitter.java:
```diff
@@ -0,0 +1,278 @@
+  public int emit(Event event) throws IOException, EventSystemException {
+    byte[] msg = event.serialize();
+    int bytesEmitted = 0;
+    synchronized (lock) {
```
I know this bit was in the original code, but do we really need to synchronize here?  emit basically just calls socket.send, which I think is threadsafe.  collectStatistics does stuff that doesn't appear threadsafe, so maybe there's little to gain in calling emit outside of a synchronized block.  one could probably also make collectStatistics threadsafe as well, but I'm just venturing further away from the intent of this PR.

nimaco commented on src/main/java/org/lwes/emitter/EmitterGroupBuilder.java:
```diff
@@ -153,71 +130,129 @@ public static EmitterGroup createGroup(Properties props, String groupName) throw
+          throw new RuntimeException(
+            String.format("Unable to parse LWES emitter group %s host config %s",
+                          groupName, host));
+        }
```
HOST can be a hostname and not just an ip address, can't it?  If so, a non-fully qualified hostname ("myserver") would not be detected by testing for ".".  PORT and TTL must be numbers, so you may be better off using them to figure out what format is specified.

kkharma replied:
updated to address this...

nimaco commented on src/main/java/org/lwes/emitter/EmitterGroupBuilder.java:
```diff
@@ -153,71 +139,138 @@ public static EmitterGroup createGroup(Properties props, String groupName) throw
+        } else if (parts.length == 4) {
+          ifaceStr = parts[0];
+          host = parts[1];
+          port = Integer.parseInt(parts[2]);
```
This parse is outside of a try-catch.  Did we want to throw a NFE out of createEmitters()?

kkharma replied:
i figured it was ok because the same is true with the previous code.

nimaco commented:
lgtm

kkharma commented:
thanks for the reviews everyone... i am going to merge these changes to master.
